src/plugins/intel_gpu/src/graph/impls/common/loop.cpp

```diff
@@ -37,63 +107,173 @@ struct loop_impl : typed_primitive_impl<loop> {
+            if (iter == 0) {
+                if (mapping.from_mem != nullptr) {
+                    mapping.from_mem->copy_from(network->get_stream(), *(mapping.initial_mem));
+                }
+            } else {
+                // In dynamic model, output memory is not defined before execution.
+                // After body network execution, replace input memory from initial_mem(external input memory) to output memory.
+                if (mapping.from_mem == nullptr) {
+                    mapping.from_mem = mapping.from_primitive->output_memory_ptr();
+                    OPENVINO_ASSERT(mapping.from_mem != nullptr, "from_mem should not be null");
+                    set_body_output_memory(network, mapping.to_primitive, mapping.from_mem);
+                }
+            }
+        } else if (mapping.type ==  loop_inst::backedge_memory_mapping::SINGLE) {
+            memory::ptr mem1 = mapping.to_primitive->output_memory_ptr();
+            if (iter == 0) {
+                mem1->copy_from(network->get_stream(), *(mapping.initial_mem));
+            } else {
+                memory::ptr mem2 = mapping.from_primitive->output_memory_ptr();
+                set_body_output_memory(network, mapping.to_primitive, std::move(mem2));
+                set_body_output_memory(network, mapping.from_primitive, std::move(mem1));
+            }
+        }
+    }
+
+    void change_memory_buffer_dynamic_model(const loop_inst::backedge_memory_mapping& mapping, network::ptr network, int64_t iter) const {
+        if (mapping.type == loop_inst::backedge_memory_mapping::CONCAT_OUTPUT) {
+            if (iter == 0) {
+                set_body_output_memory(network, mapping.to_primitive, mapping.initial_mem);
+            } else if (iter > 0) {
+                memory::ptr mem1 = mapping.to_primitive->output_memory_ptr();
+                memory::ptr mem2 = mapping.from_primitive->output_memory_ptr();
+                mem1->copy_from(network->get_stream(), *(mem2));
+            } else {
+                OPENVINO_THROW("Invalid iteration count", iter);
+            }
+        } else if (mapping.type ==  loop_inst::backedge_memory_mapping::SINGLE_SHARED) {
+            if (iter == 0) {
+                if (mapping.from_mem != nullptr) {
+                    mapping.from_mem->copy_from(network->get_stream(), *(mapping.initial_mem));
+                }
+            } else {
+                // In dynamic model, output memory is not defined before execution.
+                // After body network execution, replace input memory from initial_mem(external input memory) to output memory.
+                if (mapping.from_mem == nullptr) {
+                    mapping.from_mem = mapping.from_primitive->output_memory_ptr();
+                    OPENVINO_ASSERT(mapping.from_mem != nullptr, "from_mem should not be null");
+                    set_body_output_memory(network, mapping.to_primitive, mapping.from_mem);
```

Comment:
What is difference from change_memory_static_model?

Reply:
There are no difference between static and dynamic. Thus I unified two functions to single function.

Reply:
Please review updated code again.